# Project 06 -- Anthony Micco

**TA Help:** N/A
    
**Collaboration:** N/A

## Question 1

In [2]:
#imports
import pandas as pd
from pathlib import Path

In [3]:
#Setting the value of myyear
myyear = 1980

In [4]:
#Creating path variable
electPath = f"/anvil/projects/tdm/data/election/itcont{myyear}.txt"

In [5]:
mycolumnnames=["CMTE_ID","AMNDT_IND","RPT_TP","TRANSACTION_PGI","IMAGE_NUM","TRANSACTION_TP","ENTITY_TP","NAME","CITY","STATE","ZIP_CODE","EMPLOYER","OCCUPATION","TRANSACTION_DT","TRANSACTION_AMT","OTHER_ID","TRAN_ID","FILE_NUM","MEMO_CD","MEMO_TEXT","SUB_ID"]
mydictionarytypes = {"CMTE_ID": str, "AMNDT_IND": str, "RPT_TP": str, "TRANSACTION_PGI": str, "IMAGE_NUM": str, "TRANSACTION_TP": str, "ENTITY_TP": str, "NAME": str, "CITY": str, "STATE": str, "ZIP_CODE": str, "EMPLOYER": str, "OCCUPATION": str, "TRANSACTION_DT": str, "TRANSACTION_AMT": float, "OTHER_ID": str, "TRAN_ID": str, "FILE_NUM": str, "MEMO_CD": str, "MEMO_TEXT": str, "SUB_ID": int}
myDF = pd.read_csv(electPath, delimiter='|', names=mycolumnnames, dtype=mydictionarytypes)
myDF['TRANSACTION_DT'] = pd.to_datetime(myDF['TRANSACTION_DT'], format="%m%d%Y")

In [6]:
myDF.head()

CMTE_ID AMNDT_IND RPT_TP TRANSACTION_PGI    IMAGE_NUM TRANSACTION_TP  \
0  C00078279         A    M11               P  80031492155            22Y   
1  C00078279         A    M11             NaN  79031415137             15   
2  C00078279         A    M11             NaN  79031415137             15   
3  C00078279         A    M11             NaN  79031415137             15   
4  C00078287         A     Q1             NaN  79031231889             15   

  ENTITY_TP               NAME     CITY STATE  ...         EMPLOYER  \
0       NaN      MCKENNON, K R  MIDLAND    MI  ...              NaN   
1       NaN        OREFFICE, P  MIDLAND    MI  ...  DOW CHEMICAL CO   
2       NaN          DOWNEY, J  MIDLAND    MI  ...  DOW CHEMICAL CO   
3       NaN           BLAIR, E  MIDLAND    MI  ...  DOW CHEMICAL CO   
4       NaN  BLANCHARD, JOHN A  CHICAGO    IL  ...              NaN   

  OCCUPATION TRANSACTION_DT TRANSACTION_AMT  OTHER_ID TRAN_ID FILE_NUM  \
0        NaN     1979-10-03           400.0       NaN     NaN      NaN   
1        NaN     1979-10-26          1500.0       NaN     NaN      NaN   
2        NaN     1979-10-26           300.0       NaN     NaN      NaN   
3        NaN     1979-10-26          1000.0       NaN     NaN      NaN   
4        NaN     1979-03-20           200.0       NaN     NaN      NaN   

  MEMO_CD                       MEMO_TEXT               SUB_ID  
0     NaN  CONTRIBUTION REF TO INDIVIDUAL  3062020110011466469  
1     NaN                             NaN  3061920110000382948  
2     NaN                             NaN  3061920110000382949  
3     NaN                             NaN  3061920110000382950  
4     NaN                             NaN  3061920110000383914  

[5 rows x 21 columns]

In [7]:
myyear = 1984
electPath = f"/anvil/projects/tdm/data/election/itcont{myyear}.txt"
myDF = pd.read_csv(electPath, delimiter='|', names=mycolumnnames, dtype=mydictionarytypes)
myDF['TRANSACTION_DT'] = pd.to_datetime(myDF['TRANSACTION_DT'], format="%m%d%Y")

In [8]:
myDF.head()

CMTE_ID AMNDT_IND RPT_TP TRANSACTION_PGI    IMAGE_NUM TRANSACTION_TP  \
0  C00146845         N     MY             NaN  83012471866             15   
1  C00167973         A     MY             NaN  83012481756             15   
2  C00167973         A     MY             NaN  83012481756             15   
3  C00167973         A     MY             NaN  83012481756             15   
4  C00167973         A     MY             NaN  83012481757            15C   

  ENTITY_TP                           NAME       CITY STATE  ...  \
0       NaN             ANDREWS, STEPHEN K     INDPLS    IN  ...   
1       NaN              ZARO, CHRISTOPHER      WAYNE    PA  ...   
2       NaN    KOPROWSKI, CHRISTOPHER D MD  HAVERTOWN    PA  ...   
3       NaN  KOPROWSKI, MARY MCLAUGHLIN MD  HAVERTOWN    PA  ...   
4       NaN         PETERS, FREDERICK C II        NaN   NaN  ...   

                EMPLOYER OCCUPATION TRANSACTION_DT TRANSACTION_AMT   OTHER_ID  \
0               ATTORNEY        NaN     1983-02-28           500.0        NaN   
1  REAL ESTATE DEVELOPER        NaN     1983-01-28           500.0        NaN   
2          SELF-EMPLOYED        NaN     1983-06-29          1000.0        NaN   
3              PHYSICIAN        NaN     1983-06-29          1000.0        NaN   
4                    NaN        NaN     1983-01-03           500.0  H4PA07019   

  TRAN_ID FILE_NUM MEMO_CD MEMO_TEXT               SUB_ID  
0     NaN      NaN     NaN       NaN  3061920110001124709  
1     NaN      NaN     NaN       NaN  3061920110001112098  
2     NaN      NaN     NaN       NaN  3061920110001112099  
3     NaN      NaN     NaN       NaN  3061920110001112100  
4     NaN      NaN     NaN       NaN  3061920110001112101  

[5 rows x 21 columns]

In [9]:
myyear = 1988
electPath = f"/anvil/projects/tdm/data/election/itcont{myyear}.txt"
myDF = pd.read_csv(electPath, delimiter='|', names=mycolumnnames, dtype=mydictionarytypes)
myDF['TRANSACTION_DT'] = pd.to_datetime(myDF['TRANSACTION_DT'], format="%m%d%Y")

In [10]:
myDF.head()

CMTE_ID AMNDT_IND RPT_TP TRANSACTION_PGI    IMAGE_NUM TRANSACTION_TP  \
0  C00015263         N     MY             NaN  87013363162             15   
1  C00015263         N     MY             NaN  87013363162             15   
2  C00015263         N     MY             NaN  87013363162             15   
3  C00170506         N     MY             NaN  87013363893            15C   
4  C00170506         N     MY             NaN  87013363893            15C   

  ENTITY_TP                   NAME             CITY STATE  ... EMPLOYER  \
0       NaN           ARNOLD, CARL           MCLEAN    VA  ...      NaN   
1       NaN  HUSTON, CHARLES L III  CHESTER SPRINGS    PA  ...      NaN   
2       NaN          WALL, R DUFFY          HURLOCK    MD  ...      NaN   
3       NaN                    NaN              NaN   NaN  ...      NaN   
4       NaN                    NaN              NaN   NaN  ...      NaN   

  OCCUPATION TRANSACTION_DT TRANSACTION_AMT   OTHER_ID TRAN_ID FILE_NUM  \
0        NaN     1987-06-13           500.0        NaN     NaN      NaN   
1        NaN     1987-06-13          1000.0        NaN     NaN      NaN   
2        NaN     1987-06-30          1000.0        NaN     NaN      NaN   
3        NaN     1987-01-31          2378.0  H4MO09012     NaN      NaN   
4        NaN     1987-04-04          3044.0  H4MO09012     NaN      NaN   

  MEMO_CD MEMO_TEXT               SUB_ID  
0     NaN       NaN  3061920110001680883  
1     NaN       NaN  3061920110001680884  
2     NaN       NaN  3061920110001680885  
3     NaN       NaN  3061920110001714528  
4     NaN       NaN  3061920110001714529  

[5 rows x 21 columns]

In [13]:
def read_election_year(myyear):
    #Define the column names for the data frame
    mycolumnnames=["CMTE_ID","AMNDT_IND","RPT_TP","TRANSACTION_PGI","IMAGE_NUM","TRANSACTION_TP","ENTITY_TP","NAME","CITY","STATE","ZIP_CODE","EMPLOYER","OCCUPATION","TRANSACTION_DT","TRANSACTION_AMT","OTHER_ID","TRAN_ID","FILE_NUM","MEMO_CD","MEMO_TEXT","SUB_ID"]
    
    #Define the dictionary types for each data frame column
    mydictionarytypes = {"CMTE_ID": str, "AMNDT_IND": str, "RPT_TP": str, "TRANSACTION_PGI": str, "IMAGE_NUM": str, "TRANSACTION_TP": str, "ENTITY_TP": str, "NAME": str, "CITY": str, "STATE": str, "ZIP_CODE": str, "EMPLOYER": str, "OCCUPATION": str, "TRANSACTION_DT": str, "TRANSACTION_AMT": float, "OTHER_ID": str, "TRAN_ID": str, "FILE_NUM": str, "MEMO_CD": str, "MEMO_TEXT": str, "SUB_ID": int}
    
    #Define the path with the myyear parameter
    electPath = f"/anvil/projects/tdm/data/election/itcont{myyear}.txt"
    
    #Create the data frame
    myDF = pd.read_csv(electPath, delimiter='|', names=mycolumnnames, dtype=mydictionarytypes)
    
    #Change the TRANSACTION_DT column to be in month-day-year format
    myDF['TRANSACTION_DT'] = pd.to_datetime(myDF['TRANSACTION_DT'], format="%m%d%Y")
    
    #Show the head of the dataframe
    return myDF.head()

In [14]:
read_election_year(1996)

CMTE_ID AMNDT_IND RPT_TP TRANSACTION_PGI    IMAGE_NUM TRANSACTION_TP  \
0  C00266767         N     YE             NaN  96015864853             15   
1  C00266767         N     YE             NaN  96015864853             15   
2  C00266767         N     YE             NaN  96015864853             15   
3  C00266767         N     YE             NaN  96015864853             15   
4  C00266767         N     YE             NaN  96015864854             15   

  ENTITY_TP                 NAME                CITY STATE  ...  \
0       NaN  ALBRECHT, GARY ALAN          SOUTHFIELD    MI  ...   
1       NaN     ASPATORE, WALTER    BLOOMFIELD HILLS    MI  ...   
2       NaN       AUDETTE, FRANK    BLOOMFIELD HILLS    MI  ...   
3       NaN         BANISH, GARY  BLOOMFIELD TOWNSHP    MI  ...   
4       NaN         BARRY, PETER          BIRMINGHAM    MI  ...   

           EMPLOYER OCCUPATION TRANSACTION_DT TRANSACTION_AMT  OTHER_ID  \
0       ALBRECHT CO        NaN     1995-08-16           200.0       NaN   
1   AMHERST CAPITAL        NaN     1995-08-05           500.0       NaN   
2  AUDETTE CADILLAC        NaN     1995-08-23           200.0       NaN   
3     MERRILL LYNCH        NaN     1995-10-20           500.0       NaN   
4           RETIRED        NaN     1995-09-09           300.0       NaN   

  TRAN_ID FILE_NUM MEMO_CD MEMO_TEXT               SUB_ID  
0     NaN      NaN     NaN       NaN  3061920110004733456  
1     NaN      NaN     NaN       NaN  3061920110004733457  
2     NaN      NaN     NaN       NaN  3061920110004733458  
3     NaN      NaN     NaN       NaN  3061920110004733459  
4     NaN      NaN     NaN       NaN  3061920110004733460  

[5 rows x 21 columns]

## Question 2

In [15]:
myyear = 1980

In [16]:
electPath = f"/anvil/projects/tdm/data/election/itcont{myyear}.txt"
myDF = pd.read_csv(electPath, delimiter='|', names=mycolumnnames, dtype=mydictionarytypes)
myDF['TRANSACTION_DT'] = pd.to_datetime(myDF['TRANSACTION_DT'], format="%m%d%Y")

In [17]:
myDF.head()

CMTE_ID AMNDT_IND RPT_TP TRANSACTION_PGI    IMAGE_NUM TRANSACTION_TP  \
0  C00078279         A    M11               P  80031492155            22Y   
1  C00078279         A    M11             NaN  79031415137             15   
2  C00078279         A    M11             NaN  79031415137             15   
3  C00078279         A    M11             NaN  79031415137             15   
4  C00078287         A     Q1             NaN  79031231889             15   

  ENTITY_TP               NAME     CITY STATE  ...         EMPLOYER  \
0       NaN      MCKENNON, K R  MIDLAND    MI  ...              NaN   
1       NaN        OREFFICE, P  MIDLAND    MI  ...  DOW CHEMICAL CO   
2       NaN          DOWNEY, J  MIDLAND    MI  ...  DOW CHEMICAL CO   
3       NaN           BLAIR, E  MIDLAND    MI  ...  DOW CHEMICAL CO   
4       NaN  BLANCHARD, JOHN A  CHICAGO    IL  ...              NaN   

  OCCUPATION TRANSACTION_DT TRANSACTION_AMT  OTHER_ID TRAN_ID FILE_NUM  \
0        NaN     1979-10-03           400.0       NaN     NaN      NaN   
1        NaN     1979-10-26          1500.0       NaN     NaN      NaN   
2        NaN     1979-10-26           300.0       NaN     NaN      NaN   
3        NaN     1979-10-26          1000.0       NaN     NaN      NaN   
4        NaN     1979-03-20           200.0       NaN     NaN      NaN   

  MEMO_CD                       MEMO_TEXT               SUB_ID  
0     NaN  CONTRIBUTION REF TO INDIVIDUAL  3062020110011466469  
1     NaN                             NaN  3061920110000382948  
2     NaN                             NaN  3061920110000382949  
3     NaN                             NaN  3061920110000382950  
4     NaN                             NaN  3061920110000383914  

[5 rows x 21 columns]

In [20]:
len(myDF["CMTE_ID"].unique())

3856

In [21]:
myyear = 1984
electPath = f"/anvil/projects/tdm/data/election/itcont{myyear}.txt"
myDF = pd.read_csv(electPath, delimiter='|', names=mycolumnnames, dtype=mydictionarytypes)
myDF['TRANSACTION_DT'] = pd.to_datetime(myDF['TRANSACTION_DT'], format="%m%d%Y")
len(myDF["CMTE_ID"].unique())

3733

In [22]:
myyear = 1988
electPath = f"/anvil/projects/tdm/data/election/itcont{myyear}.txt"
myDF = pd.read_csv(electPath, delimiter='|', names=mycolumnnames, dtype=mydictionarytypes)
myDF['TRANSACTION_DT'] = pd.to_datetime(myDF['TRANSACTION_DT'], format="%m%d%Y")
len(myDF["CMTE_ID"].unique())

3995

In [30]:
def committees_function(myyear):
    #Define the column names for the data frame
    mycolumnnames=["CMTE_ID","AMNDT_IND","RPT_TP","TRANSACTION_PGI","IMAGE_NUM","TRANSACTION_TP","ENTITY_TP","NAME","CITY","STATE","ZIP_CODE","EMPLOYER","OCCUPATION","TRANSACTION_DT","TRANSACTION_AMT","OTHER_ID","TRAN_ID","FILE_NUM","MEMO_CD","MEMO_TEXT","SUB_ID"]
    
    #Define the dictionary types for each data frame column
    mydictionarytypes = {"CMTE_ID": str, "AMNDT_IND": str, "RPT_TP": str, "TRANSACTION_PGI": str, "IMAGE_NUM": str, "TRANSACTION_TP": str, "ENTITY_TP": str, "NAME": str, "CITY": str, "STATE": str, "ZIP_CODE": str, "EMPLOYER": str, "OCCUPATION": str, "TRANSACTION_DT": str, "TRANSACTION_AMT": float, "OTHER_ID": str, "TRAN_ID": str, "FILE_NUM": str, "MEMO_CD": str, "MEMO_TEXT": str, "SUB_ID": int}
    
    #Define the path with the myyear parameter
    electPath = f"/anvil/projects/tdm/data/election/itcont{myyear}.txt"
    
    #Create the data frame
    myDF = pd.read_csv(electPath, delimiter='|', names=mycolumnnames, dtype=mydictionarytypes)
    
    #Change the TRANSACTION_DT column to be in month-day-year format
    myDF['TRANSACTION_DT'] = pd.to_datetime(myDF['TRANSACTION_DT'], format="%m%d%Y")
    
    #Show the number of unique committees 
    print(f"{myyear}: {len(myDF['CMTE_ID'].unique())} unique committees")

In [32]:
committees_function(1980)
committees_function(1984)
committees_function(1988)

1980: 3856 unique committees
1984: 3733 unique committees
1988: 3995 unique committees


There are 3856 unique committees in the 1980 data frame.

There are 3733 unique committees in the 1984 data frame.

There are 3995 unique committees in the 1988 data frame.



## Question 3

In [ ]:
myyear = 1980
electPath = f"/anvil/projects/tdm/data/election/itcont{myyear}.txt"
myDF = pd.read_csv(electPath, delimiter='|', names=mycolumnnames, dtype=mydictionarytypes)
myDF['TRANSACTION_DT'] = pd.to_datetime(myDF['TRANSACTION_DT'], format="%m%d%Y")
myDF.head()

CMTE_ID AMNDT_IND RPT_TP TRANSACTION_PGI    IMAGE_NUM TRANSACTION_TP  \
0  C00078279         A    M11               P  80031492155            22Y   
1  C00078279         A    M11             NaN  79031415137             15   
2  C00078279         A    M11             NaN  79031415137             15   
3  C00078279         A    M11             NaN  79031415137             15   
4  C00078287         A     Q1             NaN  79031231889             15   

  ENTITY_TP               NAME     CITY STATE  ...         EMPLOYER  \
0       NaN      MCKENNON, K R  MIDLAND    MI  ...              NaN   
1       NaN        OREFFICE, P  MIDLAND    MI  ...  DOW CHEMICAL CO   
2       NaN          DOWNEY, J  MIDLAND    MI  ...  DOW CHEMICAL CO   
3       NaN           BLAIR, E  MIDLAND    MI  ...  DOW CHEMICAL CO   
4       NaN  BLANCHARD, JOHN A  CHICAGO    IL  ...              NaN   

  OCCUPATION TRANSACTION_DT TRANSACTION_AMT  OTHER_ID TRAN_ID FILE_NUM  \
0        NaN     1979-10-03           400.0       NaN     NaN      NaN   
1        NaN     1979-10-26          1500.0       NaN     NaN      NaN   
2        NaN     1979-10-26           300.0       NaN     NaN      NaN   
3        NaN     1979-10-26          1000.0       NaN     NaN      NaN   
4        NaN     1979-03-20           200.0       NaN     NaN      NaN   

  MEMO_CD                       MEMO_TEXT               SUB_ID  
0     NaN  CONTRIBUTION REF TO INDIVIDUAL  3062020110011466469  
1     NaN                             NaN  3061920110000382948  
2     NaN                             NaN  3061920110000382949  
3     NaN                             NaN  3061920110000382950  
4     NaN                             NaN  3061920110000383914  

[5 rows x 21 columns]

In [44]:
myDF.groupby('STATE')['TRANSACTION_AMT'].sum().sort_values().tail()

STATE
IL     9105545.0
FL    10514203.0
NY    18673301.0
CA    24085171.0
TX    24472610.0
Name: TRANSACTION_AMT, dtype: float64

In [45]:
myyear = 1984
electPath = f"/anvil/projects/tdm/data/election/itcont{myyear}.txt"
myDF = pd.read_csv(electPath, delimiter='|', names=mycolumnnames, dtype=mydictionarytypes)
myDF['TRANSACTION_DT'] = pd.to_datetime(myDF['TRANSACTION_DT'], format="%m%d%Y")

In [46]:
myDF.groupby('STATE')['TRANSACTION_AMT'].sum().sort_values().tail()

STATE
IL    10034785.0
FL    10985587.0
NY    25947339.0
CA    29265671.0
TX    36044231.0
Name: TRANSACTION_AMT, dtype: float64

In [47]:
myyear = 1988
electPath = f"/anvil/projects/tdm/data/election/itcont{myyear}.txt"
myDF = pd.read_csv(electPath, delimiter='|', names=mycolumnnames, dtype=mydictionarytypes)
myDF['TRANSACTION_DT'] = pd.to_datetime(myDF['TRANSACTION_DT'], format="%m%d%Y")

In [48]:
myDF.groupby('STATE')['TRANSACTION_AMT'].sum().sort_values().tail()

STATE
NJ    18934575.0
FL    25373272.0
TX    29076324.0
NY    49898456.0
CA    60239357.0
Name: TRANSACTION_AMT, dtype: float64

In [52]:
def top_five_states(myyear):
    #Define the column names for the data frame
    mycolumnnames=["CMTE_ID","AMNDT_IND","RPT_TP","TRANSACTION_PGI","IMAGE_NUM","TRANSACTION_TP","ENTITY_TP","NAME","CITY","STATE","ZIP_CODE","EMPLOYER","OCCUPATION","TRANSACTION_DT","TRANSACTION_AMT","OTHER_ID","TRAN_ID","FILE_NUM","MEMO_CD","MEMO_TEXT","SUB_ID"]
    
    #Define the dictionary types for each data frame column
    mydictionarytypes = {"CMTE_ID": str, "AMNDT_IND": str, "RPT_TP": str, "TRANSACTION_PGI": str, "IMAGE_NUM": str, "TRANSACTION_TP": str, "ENTITY_TP": str, "NAME": str, "CITY": str, "STATE": str, "ZIP_CODE": str, "EMPLOYER": str, "OCCUPATION": str, "TRANSACTION_DT": str, "TRANSACTION_AMT": float, "OTHER_ID": str, "TRAN_ID": str, "FILE_NUM": str, "MEMO_CD": str, "MEMO_TEXT": str, "SUB_ID": int}
    
    #Define the path with the myyear parameter
    electPath = f"/anvil/projects/tdm/data/election/itcont{myyear}.txt"
    
    #Create the data frame
    myDF = pd.read_csv(electPath, delimiter='|', names=mycolumnnames, dtype=mydictionarytypes)
    
    #Change the TRANSACTION_DT column to be in month-day-year format
    myDF['TRANSACTION_DT'] = pd.to_datetime(myDF['TRANSACTION_DT'], format="%m%d%Y")
    
    #Return the sum of transactions of the top 5 states
    return myDF.groupby('STATE')['TRANSACTION_AMT'].sum().sort_values().tail()

In [55]:
top_five_states(1980)

STATE
IL     9105545.0
FL    10514203.0
NY    18673301.0
CA    24085171.0
TX    24472610.0
Name: TRANSACTION_AMT, dtype: float64

In [57]:
top_five_states(1984)

STATE
IL    10034785.0
FL    10985587.0
NY    25947339.0
CA    29265671.0
TX    36044231.0
Name: TRANSACTION_AMT, dtype: float64

In [58]:
top_five_states(1988)

STATE
NJ    18934575.0
FL    25373272.0
TX    29076324.0
NY    49898456.0
CA    60239357.0
Name: TRANSACTION_AMT, dtype: float64

## Question 4

In [60]:
myyear = 1980
electPath = f"/anvil/projects/tdm/data/election/itcont{myyear}.txt"
myDF = pd.read_csv(electPath, delimiter='|', names=mycolumnnames, dtype=mydictionarytypes)
myDF['TRANSACTION_DT'] = pd.to_datetime(myDF['TRANSACTION_DT'], format="%m%d%Y")
myDF.head()

CMTE_ID AMNDT_IND RPT_TP TRANSACTION_PGI    IMAGE_NUM TRANSACTION_TP  \
0  C00078279         A    M11               P  80031492155            22Y   
1  C00078279         A    M11             NaN  79031415137             15   
2  C00078279         A    M11             NaN  79031415137             15   
3  C00078279         A    M11             NaN  79031415137             15   
4  C00078287         A     Q1             NaN  79031231889             15   

  ENTITY_TP               NAME     CITY STATE  ...         EMPLOYER  \
0       NaN      MCKENNON, K R  MIDLAND    MI  ...              NaN   
1       NaN        OREFFICE, P  MIDLAND    MI  ...  DOW CHEMICAL CO   
2       NaN          DOWNEY, J  MIDLAND    MI  ...  DOW CHEMICAL CO   
3       NaN           BLAIR, E  MIDLAND    MI  ...  DOW CHEMICAL CO   
4       NaN  BLANCHARD, JOHN A  CHICAGO    IL  ...              NaN   

  OCCUPATION TRANSACTION_DT TRANSACTION_AMT  OTHER_ID TRAN_ID FILE_NUM  \
0        NaN     1979-10-03           400.0       NaN     NaN      NaN   
1        NaN     1979-10-26          1500.0       NaN     NaN      NaN   
2        NaN     1979-10-26           300.0       NaN     NaN      NaN   
3        NaN     1979-10-26          1000.0       NaN     NaN      NaN   
4        NaN     1979-03-20           200.0       NaN     NaN      NaN   

  MEMO_CD                       MEMO_TEXT               SUB_ID  
0     NaN  CONTRIBUTION REF TO INDIVIDUAL  3062020110011466469  
1     NaN                             NaN  3061920110000382948  
2     NaN                             NaN  3061920110000382949  
3     NaN                             NaN  3061920110000382950  
4     NaN                             NaN  3061920110000383914  

[5 rows x 21 columns]

In [62]:
myDF.groupby('EMPLOYER')['TRANSACTION_AMT'].sum().sort_values()

EMPLOYER
N AMERICAN SYSTEMS            -2000.0
HOLIDAY INN-DISNEY EAST       -2000.0
LEASE/SALES EQMT              -1000.0
MIDWEST FABRICATION CO        -1000.0
FIRST NAT BANK OF DALTON       -500.0
                              ...    
HOUSE WIFE                  4549216.0
HOMEMAKER                   4649872.0
HOUSEWIFE                   5457943.0
ATTORNEY                    5689870.0
RETIRED                     8565354.0
Name: TRANSACTION_AMT, Length: 87347, dtype: float64

In [63]:
myyear = 1984
electPath = f"/anvil/projects/tdm/data/election/itcont{myyear}.txt"
myDF = pd.read_csv(electPath, delimiter='|', names=mycolumnnames, dtype=mydictionarytypes)
myDF['TRANSACTION_DT'] = pd.to_datetime(myDF['TRANSACTION_DT'], format="%m%d%Y")

In [64]:
myDF.groupby('EMPLOYER')['TRANSACTION_AMT'].sum().sort_values()

EMPLOYER
NSF CHECK                      -8165.0
CHARLES PANKOW INC             -7500.0
REPUBLIC HEALTH CORPORAT       -2500.0
LICHLITER, JAMESON & ASS       -2500.0
ALLTEX CONSTRUCTION, INC       -1000.0
                               ...    
HOMEMAKER                    3372784.0
SELF EMPLOYED                3869043.0
HOUSEWIFE                    6313918.0
RETIRED                      9257779.0
SELF-EMPLOYED               19761941.0
Name: TRANSACTION_AMT, Length: 70172, dtype: float64

In [65]:
myyear = 1988
electPath = f"/anvil/projects/tdm/data/election/itcont{myyear}.txt"
myDF = pd.read_csv(electPath, delimiter='|', names=mycolumnnames, dtype=mydictionarytypes)
myDF['TRANSACTION_DT'] = pd.to_datetime(myDF['TRANSACTION_DT'], format="%m%d%Y")

In [67]:
myDF.groupby('EMPLOYER')['TRANSACTION_AMT'].sum().sort_values()

EMPLOYER
NSF CONTRIBUTION           -4000.0
ERSA IMPORT & EXPORT       -2000.0
SPATOLA WINES INC          -2000.0
BRESCOME DISTR             -2000.0
GROWER FARMER              -2000.0
                           ...    
PHYSICIAN                3104940.0
RETIRED                  5865073.0
ATTORNEY                 9488872.0
HOMEMAKER                9834325.0
HOUSEWIFE               14626839.0
Name: TRANSACTION_AMT, Length: 110434, dtype: float64

In [69]:
def top_employers(myyear):
    #Define the column names for the data frame
    mycolumnnames=["CMTE_ID","AMNDT_IND","RPT_TP","TRANSACTION_PGI","IMAGE_NUM","TRANSACTION_TP","ENTITY_TP","NAME","CITY","STATE","ZIP_CODE","EMPLOYER","OCCUPATION","TRANSACTION_DT","TRANSACTION_AMT","OTHER_ID","TRAN_ID","FILE_NUM","MEMO_CD","MEMO_TEXT","SUB_ID"]
    
    #Define the dictionary types for each data frame column
    mydictionarytypes = {"CMTE_ID": str, "AMNDT_IND": str, "RPT_TP": str, "TRANSACTION_PGI": str, "IMAGE_NUM": str, "TRANSACTION_TP": str, "ENTITY_TP": str, "NAME": str, "CITY": str, "STATE": str, "ZIP_CODE": str, "EMPLOYER": str, "OCCUPATION": str, "TRANSACTION_DT": str, "TRANSACTION_AMT": float, "OTHER_ID": str, "TRAN_ID": str, "FILE_NUM": str, "MEMO_CD": str, "MEMO_TEXT": str, "SUB_ID": int}
    
    #Define the path with the myyear parameter
    electPath = f"/anvil/projects/tdm/data/election/itcont{myyear}.txt"
    
    #Create the data frame
    myDF = pd.read_csv(electPath, delimiter='|', names=mycolumnnames, dtype=mydictionarytypes)
    
    #Change the TRANSACTION_DT column to be in month-day-year format
    myDF['TRANSACTION_DT'] = pd.to_datetime(myDF['TRANSACTION_DT'], format="%m%d%Y")
    
    #Return the sum of transactions of the top 5 employers
    return myDF.groupby('EMPLOYER')['TRANSACTION_AMT'].sum().sort_values().tail()

In [71]:
top_employers(1980)

EMPLOYER
HOUSE WIFE    4549216.0
HOMEMAKER     4649872.0
HOUSEWIFE     5457943.0
ATTORNEY      5689870.0
RETIRED       8565354.0
Name: TRANSACTION_AMT, dtype: float64

In [72]:
top_employers(1984)

EMPLOYER
HOMEMAKER         3372784.0
SELF EMPLOYED     3869043.0
HOUSEWIFE         6313918.0
RETIRED           9257779.0
SELF-EMPLOYED    19761941.0
Name: TRANSACTION_AMT, dtype: float64

In [73]:
top_employers(1988)

EMPLOYER
PHYSICIAN     3104940.0
RETIRED       5865073.0
ATTORNEY      9488872.0
HOMEMAKER     9834325.0
HOUSEWIFE    14626839.0
Name: TRANSACTION_AMT, dtype: float64

## Question 5

In [74]:
myyear = 1980
electPath = f"/anvil/projects/tdm/data/election/itcont{myyear}.txt"
myDF = pd.read_csv(electPath, delimiter='|', names=mycolumnnames, dtype=mydictionarytypes)
myDF['TRANSACTION_DT'] = pd.to_datetime(myDF['TRANSACTION_DT'], format="%m%d%Y")
myDF.head()

CMTE_ID AMNDT_IND RPT_TP TRANSACTION_PGI    IMAGE_NUM TRANSACTION_TP  \
0  C00078279         A    M11               P  80031492155            22Y   
1  C00078279         A    M11             NaN  79031415137             15   
2  C00078279         A    M11             NaN  79031415137             15   
3  C00078279         A    M11             NaN  79031415137             15   
4  C00078287         A     Q1             NaN  79031231889             15   

  ENTITY_TP               NAME     CITY STATE  ...         EMPLOYER  \
0       NaN      MCKENNON, K R  MIDLAND    MI  ...              NaN   
1       NaN        OREFFICE, P  MIDLAND    MI  ...  DOW CHEMICAL CO   
2       NaN          DOWNEY, J  MIDLAND    MI  ...  DOW CHEMICAL CO   
3       NaN           BLAIR, E  MIDLAND    MI  ...  DOW CHEMICAL CO   
4       NaN  BLANCHARD, JOHN A  CHICAGO    IL  ...              NaN   

  OCCUPATION TRANSACTION_DT TRANSACTION_AMT  OTHER_ID TRAN_ID FILE_NUM  \
0        NaN     1979-10-03           400.0       NaN     NaN      NaN   
1        NaN     1979-10-26          1500.0       NaN     NaN      NaN   
2        NaN     1979-10-26           300.0       NaN     NaN      NaN   
3        NaN     1979-10-26          1000.0       NaN     NaN      NaN   
4        NaN     1979-03-20           200.0       NaN     NaN      NaN   

  MEMO_CD                       MEMO_TEXT               SUB_ID  
0     NaN  CONTRIBUTION REF TO INDIVIDUAL  3062020110011466469  
1     NaN                             NaN  3061920110000382948  
2     NaN                             NaN  3061920110000382949  
3     NaN                             NaN  3061920110000382950  
4     NaN                             NaN  3061920110000383914  

[5 rows x 21 columns]

In [76]:
myDF.groupby('TRANSACTION_DT')["TRANSACTION_AMT"].sum().sort_values()

TRANSACTION_DT
1978-01-18          4.0
1977-09-20          5.0
1977-09-19          6.0
1978-04-27         10.0
1978-01-24         10.0
                ...    
1980-06-30    1644472.0
1980-09-29    1654498.0
1980-09-30    2081495.0
1979-05-16    2691174.0
1979-07-10    4141528.0
Name: TRANSACTION_AMT, Length: 894, dtype: float64

In [78]:
myyear = 1984
electPath = f"/anvil/projects/tdm/data/election/itcont{myyear}.txt"
myDF = pd.read_csv(electPath, delimiter='|', names=mycolumnnames, dtype=mydictionarytypes)
myDF['TRANSACTION_DT'] = pd.to_datetime(myDF['TRANSACTION_DT'], format="%m%d%Y")

In [79]:
myDF.groupby('TRANSACTION_DT')["TRANSACTION_AMT"].sum().sort_values()

TRANSACTION_DT
1985-12-03         20.0
1982-03-11        100.0
1982-12-28        120.0
1982-08-27        150.0
1985-05-25        275.0
                ...    
1983-12-30    1398749.0
1984-10-01    1584482.0
1984-07-06    1651271.0
1983-09-01    1746929.0
1983-06-30    3196779.0
Name: TRANSACTION_AMT, Length: 947, dtype: float64

In [80]:
myyear = 1984
electPath = f"/anvil/projects/tdm/data/election/itcont{myyear}.txt"
myDF = pd.read_csv(electPath, delimiter='|', names=mycolumnnames, dtype=mydictionarytypes)
myDF['TRANSACTION_DT'] = pd.to_datetime(myDF['TRANSACTION_DT'], format="%m%d%Y")

In [81]:
myDF.groupby('TRANSACTION_DT')["TRANSACTION_AMT"].sum().sort_values()

TRANSACTION_DT
1985-12-03         20.0
1982-03-11        100.0
1982-12-28        120.0
1982-08-27        150.0
1985-05-25        275.0
                ...    
1983-12-30    1398749.0
1984-10-01    1584482.0
1984-07-06    1651271.0
1983-09-01    1746929.0
1983-06-30    3196779.0
Name: TRANSACTION_AMT, Length: 947, dtype: float64

In [84]:
def top_dates(myyear):
     #Define the column names for the data frame
    mycolumnnames=["CMTE_ID","AMNDT_IND","RPT_TP","TRANSACTION_PGI","IMAGE_NUM","TRANSACTION_TP","ENTITY_TP","NAME","CITY","STATE","ZIP_CODE","EMPLOYER","OCCUPATION","TRANSACTION_DT","TRANSACTION_AMT","OTHER_ID","TRAN_ID","FILE_NUM","MEMO_CD","MEMO_TEXT","SUB_ID"]
    
    #Define the dictionary types for each data frame column
    mydictionarytypes = {"CMTE_ID": str, "AMNDT_IND": str, "RPT_TP": str, "TRANSACTION_PGI": str, "IMAGE_NUM": str, "TRANSACTION_TP": str, "ENTITY_TP": str, "NAME": str, "CITY": str, "STATE": str, "ZIP_CODE": str, "EMPLOYER": str, "OCCUPATION": str, "TRANSACTION_DT": str, "TRANSACTION_AMT": float, "OTHER_ID": str, "TRAN_ID": str, "FILE_NUM": str, "MEMO_CD": str, "MEMO_TEXT": str, "SUB_ID": int}
    
    #Define the path with the myyear parameter
    electPath = f"/anvil/projects/tdm/data/election/itcont{myyear}.txt"
    
    #Create the data frame
    myDF = pd.read_csv(electPath, delimiter='|', names=mycolumnnames, dtype=mydictionarytypes)
    
    #Change the TRANSACTION_DT column to be in month-day-year format
    myDF['TRANSACTION_DT'] = pd.to_datetime(myDF['TRANSACTION_DT'], format="%m%d%Y")
    
    #Return the sum of transactions of the top 5 employers
    return myDF.groupby('TRANSACTION_DT')["TRANSACTION_AMT"].sum().sort_values().tail()

In [85]:
top_dates(1980)

TRANSACTION_DT
1980-06-30    1644472.0
1980-09-29    1654498.0
1980-09-30    2081495.0
1979-05-16    2691174.0
1979-07-10    4141528.0
Name: TRANSACTION_AMT, dtype: float64

In [86]:
top_dates(1984)

TRANSACTION_DT
1983-12-30    1398749.0
1984-10-01    1584482.0
1984-07-06    1651271.0
1983-09-01    1746929.0
1983-06-30    3196779.0
Name: TRANSACTION_AMT, dtype: float64

In [87]:
top_dates(1988)

TRANSACTION_DT
1988-08-24    2915217.0
1987-06-30    3019472.0
1988-09-30    3244376.0
1987-12-31    3420026.0
1988-06-30    3772661.0
Name: TRANSACTION_AMT, dtype: float64

## Pledge

By submitting this work I hereby pledge that this is my own, personal work. I've acknowledged in the designated place at the top of this file all sources that I used to complete said work, including but not limited to: online resources, books, and electronic communications. I've noted all collaboration with fellow students and/or TA's. I did not copy or plagiarize another's work.

> As a Boilermaker pursuing academic excellence, I pledge to be honest and true in all that I do. Accountable together – We are Purdue.